Update AttributeBehavior.php

bscheshirwork opened the pull request:
Add `$values` feature

Provide difference assigned values by attributes is user friendly.
Array of behaviors:
```php
            [
                'class' => AttributeBehavior::className(),
                'attributes' => [
                    ActiveRecord::EVENT_BEFORE_INSERT => ['number', 'name'],
                ],
                'values' => [
                    'number' => [$this, 'getNextNumber'],
                    'name' => [$this, 'getNewName']
                ]
            ],
```
instead of 
```php
            [
                'class' => AttributeBehavior::className(),
                'attributes' => [
                    ActiveRecord::EVENT_BEFORE_INSERT => 'number',
                ],
                'value' => [$this, 'getNextNumber']
            ],
           [
                'class' => AttributeBehavior::className(),
                'attributes' => [
                    ActiveRecord::EVENT_BEFORE_INSERT => 'name',
                ],
                'value' => [$this, 'getNewName']
            ],
```

| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes/no
| Fixed issues  | comma-separated list of tickets # fixed by the PR, if any

bscheshirwork committed:
Update AttributeBehavior.php

add `$values`

Kolyunya commented:
If I'm not mistaking this was rejected couple of years ago. One attribute behavior calculates one value by design.

bscheshirwork commented:
Some strange:
Multiply attributes can be change on each event, but all of this is same

Kolyunya commented:
Indeed multiple attributes can be changed but with the same value.

bscheshirwork commented:
So... I think this behavior can be more powerful. 
Because already can process multiply attributes.

SilverFire commented on framework/behaviors/AttributeBehavior.php:
```diff
@@ -80,6 +80,12 @@ class AttributeBehavior extends Behavior
     public $value;
     /**
+     * @var array an array of the values indexed by attribute name that will be assigned to the current attributes respectively by attribute name.
+     * Will be assigned if $this->value is NULL
```
<code>This property will be used only when [[value]] is \`null\`</code>

SilverFire commented on framework/behaviors/AttributeBehavior.php:
```diff
@@ -80,6 +80,12 @@ class AttributeBehavior extends Behavior
     /**
+     * @var array an array of the values indexed by attribute name that will be assigned to the current attributes respectively by attribute name.
+     * Will be assigned if $this->value is NULL
+     * The single value same at @see $value
```
I'd clarify description a bit

key - attribute name
value - corresponding value
Example:
\`\`\`php
'values' => [
    'number' => [$this, 'getNextNumber'],
    'status' => 'updated'
]
\`\`\`
@see value for possible values

SilverFire replied:
Add `@since 2.0.13`

bscheshirwork replied:
Add @since 2.0.13

its not `automatic` tag?

SilverFire commented on framework/behaviors/AttributeBehavior.php:
```diff
@@ -138,4 +147,25 @@ protected function getValue($event)
+     */
+    protected function getValueByAttribute($event, $attribute)
+    {
+        if (array_key_exists($attribute, $this->values)) {
```
`array_key_exists` will return `true`, if the value is `null`.
Function `isset()` returns `false` for `null` value, but works faster. Since value `null` should not be handled in a special way, `isset($this->values[$attribute])` should be enough

bscheshirwork replied:
So... `null` is correct value for `value` (`value` is null in current representation by default)

SilverFire replied:
So `isset` is suitable, isn't it?

bscheshirwork replied:
>Since value null should not be handled in a special way
> So isset is suitable, isn't it?

BC: The `getValue()` can be redefine with `null`-expected logic. 
like this
```
     protected function getValue($event)
     {
       if (is_null($this->value)) {
         return 'HAHAHA';
       } else  {
         return parent::getValue($event);
       }
     }
```

SilverFire replied:
HAHAHA, ok :)

bscheshirwork committed:
Update AttributeBehavior.php

update comment

SilverFire committed:
Update AttributeBehavior.php

bscheshirwork committed:
Update AttributeBehavior.php

Provide $currentAttribute feature.

Provide for [$this, 'method'].

I think this can be another better...
Comment, pls.

dynasource commented:
Im against this approach. The adding of a new array dimension makes AttributeBehavior too complex. The current state of that class already makes me have to look twice to understand the inner workings. 

I'd rather see solution in which AttributeBehavior gets added functionality + improvements on simplicity

bscheshirwork commented:
What about new behavior? Named someone like "AttributesBehavior"... This is avoid problem with BC => may be simplify.

dynasource commented:
that could be an idea. We could deprecate AttributeBehavior in favor of AttributesBehavior

Faryshta commented on framework/behaviors/AttributeBehavior.php:
```diff
@@ -80,6 +80,25 @@ class AttributeBehavior extends Behavior
+    /**
+     * @var string current processed attribute name. Can be use inside callable item of [[values]].
+     */
+    public $currentAttribute;
```
why public instead of protected?

bscheshirwork replied:
May be discussed.
Fore external anonymous function.

And... I think new behavior (prev comment) is better idea.

Faryshta commented on framework/behaviors/AttributeBehavior.php:
```diff
@@ -112,9 +131,12 @@ public function evaluateAttributes($event)
+            $attributes = (array)$this->attributes[$event->name];
             $value = $this->getValue($event);
             foreach ($attributes as $attribute) {
+                if (is_null($value)) {
```
`$value === null` is much faster

and what if i want to save a null value?

bscheshirwork replied:
If u want save null see ln183

samdark commented:
Closed in favor of #14151
